zipkin-ui/README.md

```diff
@@ -0,0 +1,34 @@
+# zipkin-ui
+
+Zipkin-UI is a single-page application that reads configuration from `/config.json`.
+
+When looking at a trace, the browser is sent to the path `/traces/{id}`. For the single-page
+app to serve that route, the server needs to forward the request to `index.html`. The same
+forwarding applies to `/dependencies` and any other routes the UI controls. 
+
+Under the scenes the JavaScript code looks at `window.location` to figure out what the
+UI should do. This is handled by a route api defined in the crossroads library.
+
+The full list of
+such routes is defined in [`resources/zipkin-ui/index-routes.json`](src/main/resources/zipkin-ui/index-routes.json).
+Routes ending with `/` should be used in prefix matching (`/foo/` should match `/foo/`, `/foo/bar`, and `/foo/bar/baz`).
+Routes not ending with `/` should be used in exact matching (`/foo` should match `/foo` and nothing else).
+
+Based on all these, there are several ways for serving zipkin-ui:
+
+ * Programatically read `index-routes.json` and set up explicit routes based on it. Example: [zipkin-web](https://github.com/openzipkin/zipkin/blob/5dec252e4c562b21bac5ac2f9d0b437d90988f79/zipkin-web/src/main/scala/com/twitter/zipkin/web/Main.scala#L92-L95).
```

Comment:
> if we know something's not there, I have a slight preference to let the server already tell me so.

The problem is that we often don't know that on beforehand on the server side. E.g. if you hit /trace/nonexistenttrace, the server would say that route exists. Except we want to show an error message, "trace not found". But in order to do that, we need to do the API call from the client, to find out that the trace indeed doesn't exist.

The problem with debugging here, is that our configuration of crossroads in zipkin-ui isn't configured to handle scenarios where it gets a location and it has no route handler for that location. We should present a generic "not found" message in the UI. In the network tab, we would see that we got a 404 from the api request.

Some initial head-scratching is unfortunately the cost of migrating to pure client-side applications... but on the other hand, what we gain in the long term, is the complete separation of frontend and backend code, so there are no "leaks" between them.